### ISSUE TITLE
custom calibration - strange colours

### ISSUE THREAD

**Issue by elstoc** [darktable-org/darktable#10482]
I have followed the instructions given by @aurelienpierre in his recent [youtube video](https://www.youtube.com/watch?v=EteAhSN9W8s) for creating custom D65 settings in the white balance module and using a color checker to create custom presets in the color calibration module. Having set both of these modules to my satisfaction, I get occasional colour issues while zooming. Sometimes, when changing a module's parameters or navigating the history stack, these strange colours persist until I change another setting or zoom in/out.

Looks like there might be some cached step that doesn't get properly invalidated. Here's an example of colour changes while zooming:

https://user-images.githubusercontent.com/9555491/143455774-47fb97a4-1aeb-46aa-abf6-2d621304f1c7.mp4

**Comment by aurelienpierre** [comment]
Now the game is to find a systematic reproducer to start investigating the cause.

**Comment by elstoc** [comment]
If it helps, the colours that appear look the same as I get if I disable color calibration (or set the illuminant to "same as pipeline"). I can quite often reproduce it by enabling color calibration, disabling it again, selecting the step before I disabled it in the history stack and then compressing history.

**Comment by elstoc** [comment]
IMO a fix is needed for 3.8 for this. I get this issue on virtually every image for which I use a custom white balance (which is most of them right now) and it means I can never tell if the image I'm seeing is the actual output from the pipe or an artifact. Sometimes just altering a random module will cause the image colours to change and the only (temporary resolution) is to come out of the darkroom and back in again.

So I'm adding it to the 3.8 milestone for now.

**Comment by elstoc** [comment]
Another example. In this case the "artifact" persisted long enough for me to create a snapshot. I can't reproduce this look by manually changing any module settings so it must be some inconsistency in the pipe. I know the white balance coefficients are passed around in the pipe. Could they be getting out of line (somehow) with what the module is doing?

![Screenshot_2021-11-27_11-52-28](https://user-images.githubusercontent.com/9555491/143680300-7b139864-f297-4941-9a9e-8712bfbd1e03.png)

Incorrect display on the left, correct display on the right.

**Comment by elstoc** [comment]
And now I've worked out how to make this happen consistently. If I change `src/iop/temperature.c` to make the population of the temperature coefficients conditional (so that they're never repopulated after the first attempt) as follows:

```
  if(!self->dev->proxy.wb_coeffs[0])
  {
    for(int k = 0; k < 4; k++)
    {
      self->dev->proxy.wb_coeffs[k] = d->coeffs[k];
    }
  }
```
Then I get the following image immediately on entering the darkroom:
![Screenshot_2021-11-27_12-07-25](https://user-images.githubusercontent.com/9555491/143680847-41c25f58-f2bf-4802-96c1-3fc76a4224f3.png)

Without this change the image looks as follows and I can only reproduce the above after fiddling around with the pipe for a while:
![Screenshot_2021-11-27_12-09-00](https://user-images.githubusercontent.com/9555491/143680856-d5de5ce9-e1ca-4b32-bf77-59f38b441f90.png)

Perhaps, then, the solution might be to also repopulate the coefficients in the `process()` function?

**Comment by paolodepetrillo** [comment]
I think this may be a race condition that occurs when white balance appears multiple times in the history stack. I was able to reproduce it by putting a delay after wb_coeffs gets set in temperature.c, to make it more likely to occur.

When dt_dev_pixelpipe_change is called, in some cases it calls commit_params on the module for every item in the history stack from oldest to newest. Every time it encounters a white balance entry, temperature.c commit_params sets dev->proxy.wb_coeffs (which is shared by all pixelpipes) to the value for that entry. So from the time it encounters the oldest white balance entry to the time it encounters the final and correct white balance, dev->proxy.wb_coeffs is set to a value that may be incorrect. If, during this time when one pipe has set wb_coeffs incorrectly, the other pipe is running process at the same time in another thread, then channelmixerrgb's process function may pick up that incorrect wb_coeffs value.

The aberrations when zooming occur when the preview pipe's output is incorrect, because the darkroom view is temporarily derived from the preview while zooming. This might be because of the race condition, or it might just be normal behavior because a module setting was changed recently and the preview pipe hasn't finished running yet for that change. The preview pipe output can be seen by enabling the overview image in upper left (shift-ctrl-n), or in the filmstrip or histogram.
